File: LeetCodeDSA/3775.py

def swap(a, b):
    temp = a 
    a = b
    b = temp

def reverseChar(s, start , end):
    while start < end: 
        swap(s[start], s[end])
        start += 1
        end -= 1
    
    
def isVowel(s):
    match s: 
        case 'a' | 'i' | 'e' | 'o' | 'u':
            return True
        case _:
            return False

def reverseWords(s):
    start = 0
    demVowel = 0
    for i in s: 
        if i != ' ':
            start += 1
            if isVowel(i):
                demVowel += 1
        else: 
            break
    
    
    i = 0
    while start < len(s):
        nb = 0
        i = start + 1
        while 1: 
            if s[i] != ' ' and i < len(s): 
                if isVowel(s[i]):
                    nb += 1
            else: 
                break
            i += 1
        if nb == demVowel:
            reverseChar(s, start + 1, i- 1)
        start = i
    
    print(s)
        
    
    
reverseWords("Huynh Pham Nhat Huy")

File: LeetCodeDSA/3775.c
#include <stdio.h>
#include <malloc.h>
#include <math.h>
#include <string.h>
#include <stdbool.h>

void swap(char *a, char *b){
    char temp = *a; 
    *a = *b; 
    *b = temp;
}

void reverseChar(char *s, int start, int end){
    while(start < end){
        swap(&(s[start]), &(s[end]));
        start++; 
        end--;
    }
}

bool isVowel(char *a){
    if( *a == 'a' || *a == 'u' || *a == 'o' || *a == 'e' || *a == 'i') return true;
    return false;
}

int main(){
    char s[] = "ae aei aeiou eio uoia";
    unsigned length = strlen(s);
    printf("%s", s);
    int chuanDem =0; 
    unsigned dem = 0; 

    for(unsigned int i =0; i < length; i++){
        if(isVowel(&(s[i]))){
            chuanDem++;
        }
        if(s[i] != ' '){
            dem++; 
        }else{ 
            break; 
        }
    }

    // printf("\n%d %d", dem, chuanDem); 


    while(s[dem] != '\0'){
        int nb =0; 
        unsigned int i = dem +1;
        while(1){
            if(s[i] != ' ' && i < length){
                if(isVowel(&(s[i]))){
                    // printf("\n%c", s[i]);
                    nb++;
                }
            }else{
                break;
            }
            i++;
        }   
        // printf("\n%c", s[i - 1]);
        if(nb == chuanDem){
            reverseChar(s, dem + 1, i - 1);
        }

        printf("\n%d", nb);
        dem = i;
    }

    printf("\n%s", s);
    return 0;
}